File: tools/install_device.py
import argparse
import contextlib
import csv
from datetime import datetime, timezone
import hashlib
import importlib
import importlib.util
import io
import json
import os
from pathlib import Path
import re
import subprocess
import sys
from types import SimpleNamespace


def require(condition, message):
    if not condition:
        raise ValueError(message)


def load_sdk(idf_path):
    parser_path = Path(idf_path) / "components/partition_table/gen_esp32part.py"
    require(parser_path.is_file(), "Set IDF_PATH or --idf-path to the ESP-IDF SDK")
    specification = importlib.util.spec_from_file_location("keyboard_partitions", parser_path)
    if specification is None or specification.loader is None:
        raise ValueError("Cannot load ESP-IDF partition parser")
    partitions = importlib.util.module_from_spec(specification)
    specification.loader.exec_module(partitions)
    esptool = importlib.import_module("esptool")
    require(esptool.__version__ in ("5.3.1", "5.4.0"),
            f"Use validated esptool 5.3.1 or 5.4.0 (found {esptool.__version__})")
    return SimpleNamespace(partitions=partitions, esptool=esptool,
                           images=importlib.import_module("esptool.bin_image"))


def image_bytes(image):
    data = Path(image["source"]).read_bytes()
    require(len(data) == image["bytes"] and hashlib.sha256(data).hexdigest() == image["sha256"],
            "Firmware file changed after validation")
    return data


def inspect_firmware(firmware, sdk):
    require(firmware.get("security") == {"secureBoot": False, "flashEncryption": False,
                                        "signedApps": False, "antiRollback": False, "httpDevelopment": True} and
            firmware["security"]["httpDevelopment"] is True,
            "A validated firmware build without security provisioning and with the HTTP owner-claim UI is required")
    images = {image["role"]: image for image in firmware["images"]}
    require(len(firmware["images"]) == 3 and set(images) == {"bootloader", "partition-table", "app"},
            "Expected exactly three firmware images")
    partition_image = images["partition-table"]
    partition_data = image_bytes(partition_image)
    require(0 < len(partition_data) <= sdk.partitions.MAX_PARTITION_LENGTH, "Unsupported partition-table image size")
    sdk.partitions.offset_part_table = partition_image["offset"]
    table = sdk.partitions.PartitionTable.from_binary(partition_data)
    table.verify()
    table.verify_size_fits(firmware["flashBytes"])
    require(not any(partition.encrypted for partition in table), "Encrypted partitions need a separate installer")
    nvs = table.find_by_name("nvs")
    if nvs is None or nvs.type != 1 or nvs.subtype != 2 or nvs.readonly:
        raise ValueError("A writable default nvs partition is required")
    require(nvs.size >= 0x3000 and nvs.size % 4096 == 0, "Unsupported NVS partition size")
    applications = [partition for partition in table if partition.type == 0]
    require(len(applications) == 1 and applications[0].subtype == 0 and
            not any(partition.type == 1 and partition.subtype == 0 for partition in table),
            "Only a single factory application without OTA metadata is supported")
    application = applications[0]
    application_write_size = ((images["app"]["bytes"] + 4095) // 4096) * 4096
    require(images["app"]["offset"] == application.offset and application_write_size <= application.size,
            "Application does not fit its declared partition")
    require(images["bootloader"]["offset"] == 0, "ESP32-S3 bootloader must start at zero")
    for role in ("bootloader", "app"):
        image = sdk.images.LoadFirmwareImage("esp32s3", image_bytes(images[role]))
        require(image.chip_id == image.ROM_LOADER.IMAGE_CHIP_ID, "Firmware header targets another chip")
        require(image.checksum == image.calculate_checksum(), "Firmware checksum mismatch")
        require(image.append_digest and image.stored_digest == image.calc_digest,
                "Firmware SHA-256 digest is missing or invalid")
    ranges = [(image["offset"], image["bytes"]) for image in images.values()] + [(nvs.offset, nvs.size)]
    previous_end = 0
    for offset, size in sorted(ranges):
        require(offset >= previous_end and offset % 4096 == 0 and size > 0, "Overlapping or unaligned flash writes")
        previous_end = offset + ((size + 4095) // 4096) * 4096
        require(previous_end <= firmware["flashBytes"], "Flash write exceeds configured capacity")
    return {"nvs": {"offset": nvs.offset, "size": nvs.size},
            "partitionTable": {"offset": partition_image["offset"], "size": 4096},
            "settings": firmware["settings"], "flashBytes": firmware["flashBytes"]}


def sync_directory(directory):
    if sys.platform == "win32":
        return
    descriptor = os.open(directory, os.O_RDONLY | os.O_DIRECTORY)
    try:
        os.fsync(descriptor)
    finally:
        os.close(descriptor)


def create_private_directory(output):
    directory = Path(output).absolute()
    repository = Path(__file__).resolve().parents[1]
    require(not directory.is_relative_to(repository), "Private installation files must be outside the repository")
    parent = directory.parent.resolve(strict=True)
    require(not parent.is_relative_to(repository), "Private output must not resolve into the repository")
    directory = parent / directory.name
    directory.mkdir(mode=0o700)
    if sys.platform == "win32":
        domain = os.environ.get("USERDOMAIN")
        username = os.environ.get("USERNAME")
        require(domain and username, "Cannot identify the Windows account for private permissions")
        permissions = subprocess.run(
            ["icacls.exe", str(directory), "/inheritance:r", "/grant:r", f"{domain}\\{username}:(OI)(CI)F"],
            capture_output=True, check=False)
        require(permissions.returncode == 0, "Cannot restrict Windows installation directory permissions")
    else:
        permissions = directory.stat()
        require(permissions.st_uid == os.getuid() and permissions.st_mode & 0o077 == 0,
                "Installation directory is not private")
    for ancestor in directory.parents:
        sync_directory(ancestor)
    return directory


def private_write(directory, name, data):
    descriptor = os.open(directory / name, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
    with os.fdopen(descriptor, "wb") as output:
        output.write(data)
        output.flush()
        os.fsync(output.fileno())
    sync_directory(directory)


def validate_identity(contents, device_id):
    rows = list(csv.DictReader(io.StringIO(contents)))
    expected = [
        ("kb_identity", "namespace", "", ""),
        ("version", "data", "u32", "1"),
        ("device_id", "data", "string", re.escape(device_id)),
        ("ap_password", "data", "string", r"[A-Za-z0-9_-]{24}"),
        ("claim_salt", "data", "hex2bin", r"[0-9a-f]{32}"),
        ("claim_hash", "data", "hex2bin", r"[0-9a-f]{64}"),
        ("claim_cost", "data", "u32", "100000"),
    ]
    require(len(rows) == len(expected), "Unexpected identity CSV records")
    for row, (key, kind, encoding, pattern) in zip(rows, expected):
        require(set(row) == {"key", "type", "encoding", "value"} and row["key"] == key and
                row["type"] == kind and row["encoding"] == encoding and
                isinstance(row["value"], str) and re.fullmatch(pattern, row["value"]) is not None,
                f"Invalid identity CSV field: {key}")


def generate_nvs(directory, device_id, size):
    source = directory / "identity.csv"
    target = directory / "identity.bin"
    validate_identity(source.read_text(encoding="utf-8"), device_id)
    require(not target.exists(), "Private NVS image already exists; do not regenerate an installation in place")
    generated = subprocess.run(
        [sys.executable, "-m", "esp_idf_nvs_partition_gen", "generate", str(source), str(target),
         hex(size), "--version", "2", "--outdir", str(directory)],
        capture_output=True, check=False)
    require(generated.returncode == 0, "ESP-IDF NVS generation failed; check the SDK environment and private CSV")
    target.chmod(0o600)
    with target.open("r+b") as generated_image:
        data = generated_image.read()
        require(len(data) == size, "NVS generator produced an unexpected image size")
        os.fsync(generated_image.fileno())
    sync_directory(directory)
    return data


def flash_capacity(value):
    require(isinstance(value, str) and re.fullmatch(r"(1|2|4|8|16|32)MB", value) is not None,
            "Unknown or unsupported flash capacity; refusing to guess")
    return int(value[:-2]) * 1048576


def install(request, sdk):
    require(request.get("execute") is True, "A device write requires --execute")
    device_id = request["deviceId"]
    require(isinstance(device_id, str) and re.fullmatch(r"[0-9a-f]{12}", device_id) is not None,
            "An expected factory base MAC is required")
    require(isinstance(request["port"], str) and request["port"] and "://" not in request["port"],
            "An explicit local serial port is required")
    require(request["baud"] in (115200, 230400, 460800, 921600), "Unsupported serial baud rate")
    identity_csv = request["identityCsv"]
    require(isinstance(identity_csv, str), "A MAC-bound identity CSV is required")
    validate_identity(identity_csv, device_id)
    firmware = request["firmware"]
    plan = inspect_firmware(firmware, sdk)
    require(flash_capacity(plan["settings"]["flash_size"]) == plan["flashBytes"], "Inconsistent flash capacity")
    directory = create_private_directory(request["directory"])
    private_write(directory, "identity.csv", identity_csv.encode("utf-8"))
    nvs_data = generate_nvs(directory, device_id, plan["nvs"]["size"])
    payloads = [(image["offset"], image_bytes(image)) for image in firmware["images"]]
    payloads.append((plan["nvs"]["offset"], nvs_data))
    payloads.sort(key=lambda item: item[0])
    manifest = {"formatVersion": 1, "deviceId": device_id, "createdAt": datetime.now(timezone.utc).isoformat(),
                "layout": plan, "replaceNvs": request.get("replaceNvs") is True,
                "images": [{"offset": offset, "bytes": len(data), "sha256": hashlib.sha256(data).hexdigest()}
                           for offset, data in payloads]}
    private_write(directory, "install-plan.json", (json.dumps(manifest, indent=2) + "\n").encode())
    backup_durability = "file-and-directory-sync"
    if sys.platform == "win32":
        backup_durability = "file-sync-only"
        print("Windows backups are verified but not guaranteed durable across host power loss; "
              "keep the host and backup storage powered.", file=sys.stderr)
    with sdk.esptool.detect_chip(port=request["port"]) as connection:
        require(connection.CHIP_NAME == "ESP32-S3", "Connected board is not an ESP32-S3")
        require(connection.secure_download_mode is False,
            "Secure download mode requires a separate installation workflow")
        secure_boot_enabled = connection.get_secure_boot_enabled()
        require(type(secure_boot_enabled) in (bool, int) and secure_boot_enabled == 0 and
            connection.get_flash_encryption_enabled() is False,
            "Secure boot or flash encryption requires a separate installation workflow")
        require(bytes(connection.read_mac(mac_type="BASE_MAC")).hex() == device_id,
            "Connected board factory base MAC does not match --device-id")
        device = sdk.esptool.run_stub(connection)
        device.change_baud(request["baud"])
        sdk.esptool.attach_flash(device)
        detected_size = sdk.esptool.detect_flash_size(device)
        detected_bytes = flash_capacity(detected_size)
        require(detected_bytes >= plan["flashBytes"], "Board flash is smaller than the firmware configuration")
        backup = sdk.esptool.read_flash(device, 0, detected_bytes, no_progress=True)
        require(isinstance(backup, bytes) and len(backup) == detected_bytes, "Incomplete flash backup")
        require(device.flash_md5sum(0, detected_bytes) == hashlib.md5(backup).hexdigest(), "Flash backup verification failed")
        private_write(directory, "flash-backup.bin", backup)
        backup_hash = hashlib.sha256(backup).hexdigest()
        require(hashlib.sha256((directory / "flash-backup.bin").read_bytes()).hexdigest() == backup_hash,
            "Saved flash backup verification failed")
        private_write(directory, "flash-backup.json", (json.dumps({"deviceId": device_id, "bytes": detected_bytes,
                  "sha256": backup_hash, "offset": 0, "durability": backup_durability}, indent=2) + "\n").encode())
        table_image = next(image for image in firmware["images"] if image["role"] == "partition-table")
        table_offset = plan["partitionTable"]["offset"]
        existing_table = backup[table_offset:table_offset + 4096]
        expected_table = image_bytes(table_image).ljust(4096, b"\xff")
        require(backup == b"\xff" * detected_bytes or existing_table == expected_table,
            "Existing partition table differs; automatic partition migration is not supported")
        nvs_offset = plan["nvs"]["offset"]
        existing_nvs = backup[nvs_offset:nvs_offset + plan["nvs"]["size"]]
        require(existing_nvs == b"\xff" * len(existing_nvs) or request.get("replaceNvs") is True,
            "NVS is not empty; --replace-nvs explicitly discards its settings and any existing ownership")
        sdk.esptool.write_flash(device, payloads, **plan["settings"], erase_all=False, force=False, no_progress=True)
        sdk.esptool.verify_flash(device, payloads, **plan["settings"])
        result = {"deviceId": device_id, "verified": True, "flashBytes": detected_bytes,
              "backupSha256": backup_hash, "backupDurability": backup_durability, "images": manifest["images"]}
        try:
            private_write(directory, "install-result.json", (json.dumps(result, indent=2) + "\n").encode())
        finally:
            sdk.esptool.reset_chip(device, "hard-reset")
    return result


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("operation", choices=["inspect", "install"])
    arguments = parser.parse_args()
    request = json.load(sys.stdin)
    with contextlib.redirect_stdout(sys.stderr):
        sdk = load_sdk(request["idfPath"])
        if arguments.operation == "inspect":
            result = inspect_firmware(request["firmware"], sdk)
        else:
            result = install(request, sdk)
    print(json.dumps(result))


if __name__ == "__main__":
    try:
        main()
    except Exception as error:
        print(f"Installer stopped: {error}", file=sys.stderr)
        sys.exit(1)

File: tools/install-device.mjs
import assert from "node:assert/strict";
import { spawn } from "node:child_process";
import { createHash } from "node:crypto";
import { constants } from "node:fs";
import { copyFile, mkdir, open, readFile, readdir, realpath, writeFile } from "node:fs/promises";
import { dirname, isAbsolute, join, relative, resolve, sep } from "node:path";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

const projectRoot = fileURLToPath(new URL("../", import.meta.url));
const sdkHelper = fileURLToPath(new URL("install_device.py", import.meta.url));
const supportedSecurity = { secureBoot: false, flashEncryption: false, signedApps: false, antiRollback: false,
  httpDevelopment: true };
const otaSecurity = { ...supportedSecurity, signedApps: true };

export function firmwareSecurity(configuration) {
  assert.equal(configuration.IDF_TARGET, "esp32s3", "Build configuration targets another chip");
  assert.equal(configuration.SECURE_BOOT, false, "Secure-boot provisioning requires a separate installer");
  assert.equal(configuration.SECURE_FLASH_ENC_ENABLED, false, "Flash-encryption provisioning requires a separate installer");
  assert.equal(configuration.KEYBOARD_HTTP_DEVELOPMENT, true, "Firmware must enable the HTTP owner-claim UI");
  if (configuration.SECURE_SIGNED_APPS_NO_SECURE_BOOT === true) {
    for (const name of ["SECURE_SIGNED_APPS_RSA_SCHEME", "SECURE_SIGNED_ON_UPDATE_NO_SECURE_BOOT",
      "SECURE_BOOT_BUILD_SIGNED_BINARIES", "BOOTLOADER_APP_ROLLBACK_ENABLE", "BOOTLOADER_WDT_ENABLE",
      "BOOTLOADER_WDT_DISABLE_IN_USER_CODE"]) assert.equal(configuration[name], true, `Missing OTA setting: ${name}`);
    assert.equal(configuration.ESPTOOLPY_FLASHSIZE, "16MB");
    for (const name of ["SECURE_BOOT_V2_ENABLED", "BOOTLOADER_APP_ANTI_ROLLBACK", "ESP_PHY_INIT_DATA_IN_PARTITION"]) {
      assert.ok(!configuration[name], `Unsupported OTA setting: ${name}`);
    }
    return { ...otaSecurity };
  }
  for (const setting of ["SECURE_BOOT_V2_ENABLED", "SECURE_BOOT_BUILD_SIGNED_BINARIES",
    "SECURE_SIGNED_APPS_NO_SECURE_BOOT", "SECURE_SIGNED_ON_UPDATE_NO_SECURE_BOOT", "BOOTLOADER_APP_ANTI_ROLLBACK"]) {
    assert.ok(configuration[setting] === undefined || configuration[setting] === false,
      `Unsupported security build configuration: ${setting}`);
  }
  return { ...supportedSecurity };
}

export function firmwareManifest(firmware) {
  if (firmware.security.signedApps) return firmware.manifest;
  return { formatVersion: 2, target: "esp32s3", security: { ...supportedSecurity },
    settings: firmware.settings, flashBytes: firmware.flashBytes,
    images: firmware.images.map(({ role, path, offset, bytes, sha256 }) => ({ role, path, offset, bytes, sha256 })) };
}

export function firmwareMetadata(flash) {
  assert.equal(flash.extra_esptool_args?.chip, "esp32s3", "Only ESP32-S3 firmware is supported");
  const settings = flash.flash_settings;
  const ota = Object.hasOwn(flash, "otadata");
  const roles = ota ? ["bootloader", "partition-table", "otadata", "app"] : ["bootloader", "partition-table", "app"];
  assert.ok(settings && ["dio", "dout", "qio", "qout"].includes(settings.flash_mode), "Invalid flash mode");
  assert.ok(["20m", "26m", "40m", "80m"].includes(settings.flash_freq), "Unsupported flash frequency");
  if (ota) assert.deepEqual(settings, { flash_mode: "dio", flash_size: "keep", flash_freq: "80m" }, "Invalid signed OTA flash settings");
  else assert.match(settings.flash_size, /^(1|2|4|8|16|32)MB$/, "A fixed supported flash size is required");
  const writeFlashArgs = ["--flash-mode", settings.flash_mode, "--flash-freq", settings.flash_freq,
    "--flash-size", settings.flash_size];
  assert.ok(Array.isArray(flash.write_flash_args) && flash.write_flash_args.length === writeFlashArgs.length,
    "Unexpected flash arguments");
  const argumentsByName = new Map();
  for (let index = 0; index < flash.write_flash_args.length; index += 2) {
    argumentsByName.set(flash.write_flash_args[index], flash.write_flash_args[index + 1]);
  }
  assert.equal(argumentsByName.size, 3, "Duplicate flash arguments");
  for (let index = 0; index < writeFlashArgs.length; index += 2) {
    assert.equal(argumentsByName.get(writeFlashArgs[index]), writeFlashArgs[index + 1], "Inconsistent flash settings");
  }
  assert.ok(flash.flash_files && Object.keys(flash.flash_files).length === roles.length,
    "Unexpected firmware image roles; no NVS or extra images allowed");
  const images = roles.map(role => {
    const image = flash[role];
    assert.ok(image && image.encrypted === "false", `Missing or encrypted ${role} image`);
    assert.match(image.offset, /^0x[0-9a-f]+$/i, "Invalid image offset");
    assert.equal(typeof image.file, "string", "Invalid image path");
    assert.ok(image.file.length > 0 && !isAbsolute(image.file) &&
      !image.file.split("/").some(part => part === ".." || part === "." || part === "") &&
      !/[\\:\x00-\x1f\x7f]/.test(image.file), "Unsafe image path");
    assert.equal(flash.flash_files[image.offset], image.file, "Inconsistent image metadata");
    const offset = Number(image.offset);
    assert.ok(Number.isSafeInteger(offset) && offset % 4096 === 0, "Image offset must be sector-aligned");
    return { role, path: image.file, offset };
  }).sort((left, right) => left.offset - right.offset);
  assert.equal(new Set(images.map(image => image.offset)).size, roles.length, "Duplicate image offsets");
  assert.equal(new Set(images.map(image => image.path)).size, roles.length, "Duplicate image paths");
  assert.equal(images[0].role, "bootloader", "Bootloader must be the first image");
  assert.equal(images[0].offset, 0, "ESP32-S3 bootloader must start at zero");
  return { settings: { flash_mode: settings.flash_mode, flash_size: settings.flash_size, flash_freq: settings.flash_freq },
    flashBytes: ota ? 0x1000000 : Number.parseInt(settings.flash_size, 10) * 1048576,
    writeFlashArgs, images };
}

export async function loadFirmware(directory) {
  const root = await realpath(resolve(directory));
  const flash = JSON.parse(await readFile(join(root, "flasher_args.json"), "utf8"));
  const metadata = firmwareMetadata(flash);
  let previousEnd = 0;
  const images = [];
  for (const image of metadata.images) {
    const source = await realpath(resolve(root, image.path));
    const contained = relative(root, source);
    assert.ok(contained !== ".." && !contained.startsWith(`..${sep}`) && !isAbsolute(contained),
      "Image resolves outside the firmware directory");
    const data = await readFile(source);
    assert.ok(data.length > 0, `Empty image: ${image.path}`);
    const end = image.offset + Math.ceil(data.length / 4096) * 4096;
    assert.ok(image.offset >= previousEnd && end <= metadata.flashBytes,
      `Overlapping or oversized image: ${image.path}`);
    previousEnd = end;
    images.push({ ...image, source, bytes: data.length, sha256: createHash("sha256").update(data).digest("hex") });
  }
  const firmware = { ...metadata, root, images, security: { ...supportedSecurity } };
  if (Object.hasOwn(flash, "otadata")) {
    const manifest = JSON.parse(await readFile(join(root, "firmware-manifest.json"), "utf8"));
    const signature = await readFile(join(root, "firmware-manifest.sig"));
    assert.equal(manifest.formatVersion, 3, "Unsupported OTA manifest schema");
    assert.equal(manifest.artifact, "keyboard-install");
    assert.equal(manifest.target, "esp32s3");
    assert.deepEqual(manifest.security, otaSecurity);
    assert.deepEqual(manifest.settings, metadata.settings);
    assert.equal(manifest.flashBytes, metadata.flashBytes);
    assert.deepEqual(manifest.images, images.map(({ role, path, offset, bytes, sha256 }) => ({ role, path, offset, bytes, sha256 })));
    assert.equal(signature.length, 384, "Invalid RSA-3072 manifest signature length");
    return { ...firmware, security: { ...otaSecurity }, manifest, manifestSignature: signature.toString("hex") };
  }
  let configuration;
  try {
    configuration = await readFile(join(root, "config/sdkconfig.json"), "utf8");
  } catch (error) {
    if (error.code !== "ENOENT") throw error;
  }
  if (configuration !== undefined) {
    firmware.security = firmwareSecurity(JSON.parse(configuration));
  } else {
    const manifest = JSON.parse(await readFile(join(root, "firmware-manifest.json"), "utf8"));
    assert.deepEqual(manifest, firmwareManifest(firmware), "Firmware manifest is unsafe or does not match the images and settings");
  }
  return firmware;
}

export function installOptions(args, environment = process.env) {
  const { values } = parseArgs({ args, allowPositionals: false, options: {
    firmware: { type: "string", default: join(projectRoot, "build") },
    "idf-path": { type: "string", default: environment.IDF_PATH },
    python: { type: "string", default: environment.PYTHON || "python" },
    "device-id": { type: "string" },
    output: { type: "string" },
    port: { type: "string" },
    baud: { type: "string", default: "460800" },
    execute: { type: "boolean", default: false },
    "write-manifest": { type: "boolean", default: false },
    "replace-nvs": { type: "boolean", default: false },
    "reset-layout": { type: "boolean", default: false },
    "verification-key": { type: "string" },
    "allow-test-firmware": { type: "boolean", default: false },
    help: { type: "boolean", short: "h", default: false },
  } });
  if (values.help) return { help: true };
  assert.ok(values["idf-path"], "Open an ESP-IDF terminal or supply --idf-path and --python");
  assert.ok(["115200", "230400", "460800", "921600"].includes(values.baud), "Unsupported serial baud rate");
  assert.ok(!values["replace-nvs"] || values.execute, "--replace-nvs requires --execute");
  assert.ok(!values["reset-layout"] || values.execute, "--reset-layout requires --execute");
  assert.ok(!(values["reset-layout"] && values["replace-nvs"]), "Choose full layout reset or legacy NVS replacement, not both");
  assert.ok(!values["write-manifest"] || !values.execute, "--write-manifest is an offline build operation");
  if (values.execute) {
    assert.match(values["device-id"] ?? "", /^[0-9a-f]{12}$/i, "--execute requires --device-id: twelve factory base MAC hex digits");
    assert.ok(values.output, "--execute requires a new private --output directory outside the repository");
    assert.ok(values.port && !values.port.includes("://"), "--execute requires an explicit local --port");
  }
  return { firmware: resolve(values.firmware), idfPath: resolve(values["idf-path"]), python: values.python,
    deviceId: values["device-id"]?.toLowerCase(), output: values.output && resolve(values.output),
    port: values.port, baud: Number(values.baud), execute: values.execute, replaceNvs: values["replace-nvs"],
    writeManifest: values["write-manifest"], resetLayout: values["reset-layout"],
    verificationKey: values["verification-key"] && resolve(values["verification-key"]),
    allowTestFirmware: values["allow-test-firmware"] };
}

function runSdk(options, operation, request) {
  return new Promise((resolveResult, reject) => {
    const child = spawn(options.python, [sdkHelper, operation], {
      stdio: ["pipe", "pipe", "pipe"], env: { ...process.env, PYTHONUNBUFFERED: "1" },
    });
    let output = "";
    child.stdout.setEncoding("utf8");
    child.stdout.on("data", chunk => { output += chunk; });
    child.stderr.on("data", chunk => { process.stderr.write(chunk); });
    child.on("error", reject);
    child.stdin.on("error", error => { if (error.code !== "EPIPE") reject(error); });
    child.on("close", code => {
      if (code !== 0) {
        reject(new Error(`ESP-IDF ${operation} failed; review the diagnostic above. No automatic retry was attempted.`));
        return;
      }
      try {
        resolveResult(JSON.parse(output));
      } catch {
        reject(new Error("ESP-IDF helper returned an invalid result; do not assume the installation succeeded"));
      }
    });
    child.stdin.end(JSON.stringify(request));
  });
}

export async function syncPrivateDirectory(directory) {
  for (const entry of await readdir(directory, { withFileTypes: true })) {
    const path = join(directory, entry.name);
    if (entry.isDirectory()) {
      await syncPrivateDirectory(path);
    } else {
      assert.ok(entry.isFile(), "Private installation output must not contain symlinks or special files");
      const file = await open(path, "r+");
      try {
        await file.sync();
      } finally {
        await file.close();
      }
    }
  }
  if (process.platform !== "win32") {
    const parent = await open(directory, "r");
    try {
      await parent.sync();
    } finally {
      await parent.close();
    }
  }
}

async function snapshotFirmware(firmware, directory) {
  const destination = join(directory, "firmware");
  await mkdir(destination, { mode: 0o700 });
  await copyFile(join(firmware.root, "flasher_args.json"), join(destination, "flasher_args.json"), constants.COPYFILE_EXCL);
  for (const image of firmware.images) {
    const target = join(destination, image.path);
    await mkdir(dirname(target), { recursive: true, mode: 0o700 });
    await copyFile(image.source, target, constants.COPYFILE_EXCL);
  }
  await writeFile(join(destination, "firmware-manifest.json"), `${JSON.stringify(firmwareManifest(firmware), null, 2)}\n`,
    { flag: "wx", mode: 0o600 });
  if (firmware.security.signedApps) {
    await writeFile(join(destination, "firmware-manifest.sig"), Buffer.from(firmware.manifestSignature, "hex"), { flag: "wx", mode: 0o600 });
  }
  const snapshot = await loadFirmware(destination);
  assert.deepEqual(snapshot.settings, firmware.settings, "Flash settings changed while creating the private package");
  const fingerprint = images => images.map(({ source, ...image }) => image);
  assert.deepEqual(fingerprint(snapshot.images), fingerprint(firmware.images), "Firmware changed while creating the private package");
  const args = [snapshot.writeFlashArgs.join(" "),
    ...snapshot.images.map(image => `0x${image.offset.toString(16)} ${JSON.stringify(image.path)}`), ""].join("\n");
  await writeFile(join(destination, "flash_args"), args, { flag: "wx", mode: 0o600 });
  return snapshot;
}

export async function runInstaller(args, dependencies = {}) {
  const options = installOptions(args);
  const log = dependencies.log ?? console.log;
  if (options.help) {
    log(`Usage: node tools/install-device.mjs [--firmware <build-or-extracted-artifact-directory>]
       [--idf-path <ESP-IDF-directory>] [--python <ESP-IDF-python>]

Without --execute: validate firmware and display its NVS layout offline. No secrets or device operations.
--write-manifest records a credential-free build manifest for distributing separate firmware images.
To install: add --execute --port <COMx> --device-id <12-hex-factory-base-MAC> --output <new-private-directory>
Optional: --baud <115200|230400|460800|921600> (default 460800).
--replace-nvs explicitly discards existing NVS settings and ownership after a verified full-flash backup.
OTA builds: --verification-key <independently trusted public PEM> is required, including offline checks.
To install an OTA build: also pass --reset-layout to confirm a full erase and fresh provisioning.
--allow-test-firmware explicitly permits installation of a test-key build. Never use a candidate bundle as its own trust root.
Legacy installations retain their existing partition guards; OTA clean installation never reads/converts old settings.
Private output includes a firmware snapshot, fresh identity, setup card, and installation verification records.
Only legacy factory installations include a full-flash backup. OTA --reset-layout creates no old-flash backup.
Never upload private output to Git or CI.`);
    return { mode: "help" };
  }
  const sdk = dependencies.sdk ?? ((operation, request) => runSdk(options, operation, request));
  const firmware = await loadFirmware(options.firmware);
  let verificationKey;
  if (firmware.security.signedApps) {
    assert.ok(options.verificationKey, "OTA verification requires an independently trusted --verification-key public PEM");
    verificationKey = await readFile(options.verificationKey, "utf8");
    assert.ok(/-----BEGIN (?:RSA )?PUBLIC KEY-----/.test(verificationKey) && !verificationKey.includes("PRIVATE KEY"),
      "Supply a public verification key, not a private signing key");
    if (options.execute) {
      assert.ok(options.resetLayout, "OTA installation requires --reset-layout: full erase and fresh provisioning");
      const keyLocation = relative(firmware.root, await realpath(options.verificationKey));
      assert.ok(keyLocation === ".." || keyLocation.startsWith(`..${sep}`) || isAbsolute(keyLocation),
        "The trusted verification key must be obtained independently, outside the candidate firmware directory");
    }
  } else assert.ok(!options.resetLayout, "Layout reset requires a validated OTA build");
  const plan = await sdk("inspect", { firmware, idfPath: options.idfPath, verificationKey });
  log(`Validated ESP32-S3 firmware: ${firmware.settings.flash_size}, ${firmware.settings.flash_mode}, ${firmware.settings.flash_freq}.`);
  for (const image of firmware.images) log(`0x${image.offset.toString(16)}  ${image.path}  ${image.bytes} bytes`);
  log(`NVS identity: offset 0x${plan.nvs.offset.toString(16)}, size 0x${plan.nvs.size.toString(16)} (from the partition table).`);
  if (!options.execute) {
    if (options.writeManifest) {
      await writeFile(join(firmware.root, "firmware-manifest.json"), `${JSON.stringify(firmwareManifest(firmware), null, 2)}\n`);
      log("Wrote firmware-manifest.json with image SHA-256 hashes and supported build-security settings. No credentials included.");
    }
    log("Offline check only. No credentials generated, serial connection, flash write, erase, or eFuse operation.");
    return { mode: "check", plan };
  }
  if (options.replaceNvs) log("NVS replacement requested: existing Wi-Fi settings and ownership will be discarded after backup.");
  if (options.resetLayout) log("Full layout reset requested: the board will be erased and provisioned with new credentials.");
  const prepareIdentity = dependencies.writeIdentity ?? (await import("./provision-device.mjs")).writeIdentity;
  const directory = await prepareIdentity(options.deviceId, options.output, firmware.security.signedApps ? 10 : 100000);
  log(`Private installation files: ${directory}. Credentials are not printed.`);
  try {
    const snapshot = await snapshotFirmware(firmware, directory);
    await (dependencies.syncPrivateDirectory ?? syncPrivateDirectory)(directory);
    log(options.resetLayout ? "Checking the board before full erase and fresh installation." :
      "Checking the board and backing up flash before one combined firmware and NVS write.");
    const result = await sdk("install", { firmware: snapshot, idfPath: options.idfPath,
      directory: join(directory, "installation"), identityCsv: await readFile(join(directory, "identity.csv"), "utf8"),
      deviceId: options.deviceId, port: options.port, baud: options.baud, execute: true, replaceNvs: options.replaceNvs,
      resetLayout: options.resetLayout, verificationKey, allowTestFirmware: options.allowTestFirmware });
    assert.equal(result.verified, true, "The helper did not confirm write verification");
    log("Firmware and NVS verified on flash. Keep the private installation files and setup card; runtime acceptance is still required.");
    return { mode: "install", directory, result };
  } catch (error) {
    log("Installation stopped. Private files are retained; a write failure may leave partial firmware or NVS. Do not regenerate credentials to retry.");
    throw error;
  }
}

if (process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  try {
    await runInstaller(process.argv.slice(2));
  } catch (error) {
    console.error(error.message);
    process.exitCode = 1;
  }
}